appliance/mariadb/process_test.go

```diff
@@ -0,0 +1,492 @@
+	Delay: 200 * time.Millisecond,
+}
+
+func assertDownstream(c *C, db *sql.DB, n int) {
+	var row struct {
+		ServerID int64
+		Host     string
+		Port     int64
+		MasterID int64
+	}
+	err := db.QueryRow("SHOW SLAVE HOSTS").Scan(
+		&row.ServerID,
+		&row.Host,
+		&row.Port,
+		&row.MasterID,
+	)
+	c.Assert(err, IsNil)
+	c.Assert(row.Host, Equals, fmt.Sprintf("node%d", n))
+}
+
+func waitRow(c *C, db *sql.DB, n int) {
+	var res int64
+	err := queryAttempts.Run(func() error {
+		return db.QueryRow(fmt.Sprintf("SELECT id FROM test WHERE id = %d", n)).Scan(&res)
+	})
+	c.Assert(err, IsNil)
+}
+
+func createTable(c *C, db *sql.DB) {
+	_, err := db.Exec("CREATE TABLE test (id bigint PRIMARY KEY)")
+	c.Assert(err, IsNil)
+	insertRow(c, db, 1)
+}
+
+func insertRow(c *C, db *sql.DB, n int) {
+	_, err := db.Exec(fmt.Sprintf("INSERT INTO test (id) VALUES (%d)", n))
+	c.Assert(err, IsNil)
+}
+
+func waitReadWrite(c *C, db *sql.DB) {
+	var readOnly string
+	err := queryAttempts.Run(func() error {
+		if err := db.QueryRow("SELECT @@read_only").Scan(&readOnly); err != nil {
+			return err
+		}
+		if readOnly == "0" {
+			return nil
+		}
+		return fmt.Errorf("transaction readonly is %q", readOnly)
+	})
+	c.Assert(err, IsNil)
+}
+
+var syncAttempts = attempt.Strategy{
+	Min:   5,
+	Total: 30 * time.Second,
+	Delay: 200 * time.Millisecond,
+}
+
+func waitReplSync(c *C, p *Process, n int) {
+	id := fmt.Sprintf("node%d", n)
+	err := syncAttempts.Run(func() error {
+		info, err := p.Info()
+		if err != nil {
+			return err
+		}
+		if info.SyncedDownstream == nil || info.SyncedDownstream.ID != id {
+			return errors.New("downstream not synced")
+		}
+		return nil
+	})
+	c.Assert(err, IsNil, Commentf("up:%s down:%s", p.ID, id))
+}
+
+func (MariaDBSuite) TestIntegration_TwoNodeSync(c *C) {
+	// Start a primary.
+	node1 := NewTestProcess(c, 1)
+	err := node1.Reconfigure(Config(state.RolePrimary, 0, 2))
+	c.Assert(err, IsNil)
+	c.Assert(node1.Start(), IsNil)
+	defer node1.Stop()
+
+	srv1 := NewHTTPServer(c, node1, 1)
+	defer srv1.Close()
+
+	// Start a sync
+	node2 := NewTestProcess(c, 2)
+	err = node2.Reconfigure(Config(state.RoleSync, 1, 0))
+	c.Assert(err, IsNil)
+	c.Assert(node2.Start(), IsNil)
+	defer node2.Stop()
+
+	srv2 := NewHTTPServer(c, node2, 2)
+	defer srv2.Close()
+
+	// check it catches up
+	waitReplSync(c, node1, 2)
+
+	// Write to the master.
+	db1 := connect(c, node1, "mysql")
+	defer db1.Close()
+	_, err = db1.Exec("CREATE DATABASE foo")
+	c.Assert(err, IsNil)
+	_, err = db1.Exec("use foo")
+	c.Assert(err, IsNil)
+	_, err = db1.Exec("CREATE TABLE bar (value VARCHAR(255))")
+	c.Assert(err, IsNil)
+	_, err = db1.Exec("INSERT INTO bar (value) VALUES ('baz')")
+	c.Assert(err, IsNil)
+
+	waitReplSync(c, node1, 2)
+
+	time.Sleep(1 * time.Second)
+
+	// Read from the slave.
+	var value string
+	db2 := connect(c, node2, "foo")
+	defer db2.Close()
+	c.Assert(db2.QueryRow("SELECT value FROM bar").Scan(&value), IsNil)
+	c.Assert(value, Equals, "baz")
```

Comment:
@josephglanville This section here was previously querying `node1` instead of `node2` so it was always succeeding. I fixed that in cd4c3de but now I'm getting an error when I query it immediately (but it works if I add a `100ms` sleep) which makes me think it's not being synchronously replicated. I tried digging into it further but I'm not seeing why it's not running semi-sync.